Sources/DangerDependenciesResolver/PackageDataProvider.swift

```diff
@@ -0,0 +1,117 @@
+import DangerShellExecutor
+import Foundation
+import Logger
+import Version
+
+protocol PackageDataProviding {
+    func nameOfPackage(at url: URL, temporaryFolder: String) throws -> String
+    func latestMajorVersionForPackage(at url: URL) throws -> Int
+    func resolvePinnedPackages(generatedFolder: String) throws -> [Package.Pinned]
+}
+
+struct PackageDataProvider: PackageDataProviding {
+    let fileReader: FileReading
+    let logger: Logger
+    let executor: ShellExecuting
+
+    init(logger: Logger,
+         fileReader: FileReading = FileReader(),
+         executor: ShellExecuting = ShellExecutor()) {
+        self.fileReader = fileReader
+        self.logger = logger
+        self.executor = executor
+    }
+
+    enum Errors: Error {
+        case failedToResolveName(URL)
+        case failedToReadPackageFile(String)
+        case failedToResolveLatestVersion(URL)
+    }
+
+    func nameOfPackage(at url: URL, temporaryFolder: String) throws -> String {
+        do {
+            guard !url.isForRemoteRepository else {
+                return try nameOfRemotePackage(at: url, temporaryFolder: temporaryFolder)
+            }
+
+            let folder = url.absoluteString
+            return try nameOfPackage(in: folder)
+        } catch {
+            throw Errors.failedToResolveName(url)
+        }
+    }
+
+    func latestMajorVersionForPackage(at url: URL) throws -> Int {
+        guard let releases = try? versions(for: url),
+            let latestVersion = releases.sorted().last else {
+            throw Errors.failedToResolveLatestVersion(url)
+        }
+
+        logger.logInfo("Using \(url.absoluteString) latest major:\(latestVersion.major)")
+
+        return latestVersion.major
+    }
+
+    func resolvePinnedPackages(generatedFolder: String) throws -> [Package.Pinned] {
+        struct ResolvedPackagesState: Decodable {
```

Comment:
<!--
  0 failure: 
  1 warning:  Types should be n...
  
  
  DangerID: danger-id-Danger;
  File: Sources/DangerDependenciesResolver/PackageDataProvider.swift;
  Line: 56;
-->

- :warning: Types should be nested at most 1 level deep (`nesting`)